### ISSUE TITLE
Suggestion: Graduall typing using soft Interfaces

### ISSUE THREAD

**Issue by ofirgeller** [microsoft/TypeScript#211]
#### tl;dr

I want to be able to declare that an object implements an interface, but still have it act like an any when referencing a property not defined by the interface
#### Use case:

in angular controllers get injected with a scope object that lets you bind var between the view and model. the scope has some methods defined on it like $watch, $on, $apply and more.  it's nice to be able to type the scope as NG.IScope to get all the regular benefits of typing. however, on every controller there will be a different set of properties set on the scope object. like
$scope.people = [{name:...},{},{},{}]

And so on, the current way of doing this is extending the IScope interface, but this gives close to zero benefit(no html compile time errors). and has a high cost in writing and maintenance. it is the reason why we are now just living the scope as <any>.
#### Result I am looking for

After defining scope , I want this to happen

$scope.$watch(124) // error, we know $scope has a function called $watch and we know it's not accepting a number as the first parameter.

$scope.people = [1,2,2] // no error, we don't know what people is so the compiler lets it slide. maybe with some optional warning.
#### Syntax option

We could re use the 'implements' keyword in a new context, in declaring the var or parameter. like so:

function PeopleCtrl ($scope implements NG.IScope){
}

This means i'm telling the compiler "scope **implements but is not limited to NG.IScope**".

If this is a bad idea, I will love to hear why also if you have a better way to solve the use case let me know.
